ansible/roles/test/tasks/pfc_wd/functional_test/functional_test.yml

```diff
@@ -0,0 +1,299 @@
+
+- name: Generate config file for drop action.
+  vars:
+      pfc_wd_interface_list: "{{ pfc_wd_test_port }}"
+      pfc_wd_action: drop
+      pfc_wd_detection_time: "{{ pfc_wd_poll_time }}"
+      pfc_wd_restoration_time: "{{ pfc_wd_penalty }}"
+      pfc_wd_op: SET
+  template:
+      src: "{{ pfc_wd_template }}"
+      dest: "{{ tests_location }}/{{testname}}/pfc_wd_drp_action.json"
+  connection: local
+  become: false
+
+- name: Generate config file for forward action.
+  vars:
+      pfc_wd_interface_list: "{{ pfc_wd_test_port }}"
+      pfc_wd_action: forward
+      pfc_wd_detection_time: "{{ pfc_wd_poll_time }}"
+      pfc_wd_restoration_time: "{{ pfc_wd_penalty }}"
+      pfc_wd_op: SET
+  template:
+      src: "{{ pfc_wd_template }}"
+      dest: "{{ tests_location }}/{{testname}}/pfc_wd_fwd_action.json"
+  connection: local
+  become: false
+
+- name: Generate config file for config removal.
+  vars:
+      pfc_wd_interface_list: "{{ pfc_wd_test_port }}"
+      pfc_wd_action: drop
+      pfc_wd_detection_time: "{{ pfc_wd_poll_time }}"
+      pfc_wd_restoration_time: "{{ pfc_wd_penalty }}"
+      pfc_wd_op: DEL
+  template:
+      src: "{{ pfc_wd_template }}"
+      dest: "{{ tests_location }}/{{testname}}/pfc_wd_del_action.json"
+  connection: local
+  become: false
+
+- name: Initialize config test
+  include: roles/test/tasks/init_config_test.yml
+
+- name: copy the test to ptf container
+  copy: src=roles/test/files/ptftests dest=/root
+  delegate_to: "{{ ptf_host }}"
+
+# Verify proper function of drop action
+- block:
+    - name: Apply drop config to {{ pfc_wd_test_port }}.
+      vars:
+        config_file: pfc_wd_drp_action.json
+        errors_expected: false
+        run_cleanup: false
+      include: roles/test/tasks/run_config_test.yml
+
+    - set_fact:
+        test_expect_file: "expect_pfc_wd_detect"
+
+    - name: Initialize loganalyzer
+      include: roles/test/files/tools/loganalyzer/loganalyzer_init.yml
+
+    - name: Generate PFC storm on fanout switch
+      action: apswitch template=pfc_storm.j2
+      args:
+        host: "{{peer_device}}"
+        login: "{{peer_login}}"
+      connection: switch
+
+    - name: Check if logs contain message that PFC WD detected deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_analyze.yml
+
+    - name: Check if logs contain message that PFC WD detected deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_end.yml
+
+    - name: "Send 10K packets to {{pfc_wd_test_port}}"
+      include: roles/test/tasks/ptf_runner.yml
+      vars:
+        ptf_test_name: PFC WD test
+        ptf_test_dir: ptftests
+        ptf_test_path: pfc_wd.PfcWdTest
+        ptf_platform: remote
+        ptf_test_params:
+          - testbed_type='{{testbed_type}}'
+          - router_mac='{{ansible_Ethernet0['macaddress']}}'
+          - queue_index='{{pfc_queue_index}}'
+          - pkt_count='10000'
+          - port_src='{{pfc_wd_test_port_id}}'
```

Comment:
Is port_src pfc_wd_rx_port_id or test_port_id? I think you are going to send the data packets to rx_port_id?

Reply:
It's the other way. Rx port on ptf docker is Tx port on DUT.

Reply:
I am lost here. Could you elaborate a little bit more? I was thinking the packet goes to  rx_port on DUT and DUT will try to forward the packet to test_port neighbor via test_port. If this is the case, why the src_port equals to test_port? Let me know if I miss something.